### PR TITLE
Improve the REST API error handling

### LINKED ISSUE #1449: [BUG]: Confusing error message while attempting to connect to ZenML server
### Contact Details [Optional]

_No response_

### System Information

ZENML_LOCAL_VERSION: 0.35.1
ZENML_SERVER_VERSION: 0.35.1
ZENML_SERVER_DATABASE: sqlite
ZENML_SERVER_DEPLOYMENT_TYPE: other
ZENML_CONFIG_DIR: /Users/thomaskwok/Library/Application Support/zenml
ZENML_LOCAL_STORE_DIR: /Users/thomaskwok/Library/Application Support/zenml/local_stores
ZENML_SERVER_URL: sqlite:////Users/thomaskwok/Library/Application Support/zenml/local_stores/default_zen_store/zenml.db
PYTHON_VERSION: 3.10.9
ENVIRONMENT: native
SYSTEM_INFO: {'os': 'mac', 'mac_version': '12.6'}
ACTIVE_WORKSPACE: default
ACTIVE_STACK: stack_1
ACTIVE_USER: default
TELEMETRY_STATUS: enabled
ANALYTICS_CLIENT_ID: e4fc518b-4841-440b-ae6b-6e5c0758aa24
ANALYTICS_USER_ID: 5b978a0b-c9ca-4879-842a-6e5f3f111f74
ANALYTICS_SERVER_ID: e4fc518b-4841-440b-ae6b-6e5c0758aa24
INTEGRATIONS: ['kaniko', 'kubeflow', 'kubernetes', 'pillow', 'pytorch', 'scipy', 'sklearn']
PACKAGES: {'deprecated': '1.2.13', 'flask': '2.2.3', 'gitpython': '3.1.31', 'jinja2': '3.1.2', 'mako': '1.2.4', 'markdown': '3.4.3', 'markupsafe': '2.1.2', 'pillow': '9.4.0', 'pyjwt': '2.6.0', 'pymysql': '1.0.3', 'pyyaml': '5.4.1', 
'pygments': '2.14.0', 'sqlalchemy': '1.4.41', 'sqlalchemy-utils': '0.38.3', 'send2trash': '1.8.0', 'werkzeug': '2.2.3', 'absl-py': '1.4.0', 'aiohttp': '3.8.4', 'aiosignal': '1.3.1', 'alembic': '1.8.1', 'analytics-python': 
'1.4.post1', 'anyio': '3.6.2', 'appnope': '0.1.3', 'argon2-cffi': '21.3.0', 'argon2-cffi-bindings': '21.2.0', 'asgiref': '3.6.0', 'asttokens': '2.2.1', 'async-timeout': '4.0.2', 'attrs': '22.2.0', 'backcall': '0.2.0', 'backoff': 
'1.10.0', 'bcrypt': '4.0.1', 'beautifulsoup4': '4.12.0', 'bleach': '6.0.0', 'cachetools': '5.3.0', 'certifi': '2022.12.7', 'cffi': '1.15.1', 'charset-normalizer': '3.1.0', 'class-resolver': '0.3.10', 'click': '8.1.3', 
'click-params': '0.3.0', 'click-spinner': '0.1.10', 'cloudpickle': '2.2.1', 'colorama': '0.4.6', 'comm': '0.1.3', 'commonmark': '0.9.1', 'contourpy': '1.0.7', 'cryptography': '40.0.1', 'cycler': '0.11.0', 'databricks-cli': '0.17.6',
'debugpy': '1.6.6', 'decorator': '5.1.1', 'defusedxml': '0.7.1', 'distro': '1.8.0', 'docker': '6.0.1', 'docstring-parser': '0.15', 'dyadctl': '2.18.0', 'ecdsa': '0.18.0', 'entrypoints': '0.4', 'exceptiongroup': '1.1.1', 'executing':
'1.2.0', 'fastapi': '0.75.2', 'fastapi-utils': '0.2.1', 'fastjsonschema': '2.16.3', 'fire': '0.5.0', 'fonttools': '4.39.2', 'frozenlist': '1.3.3', 'fsspec': '2023.3.0', 'gcsfs': '2023.3.0', 'gitdb': '4.0.10', 'google-api-core': 
'2.11.0', 'google-api-python-client': '1.12.11', 'google-auth': '2.16.3', 'google-auth-httplib2': '0.1.0', 'google-auth-oauthlib': '1.0.0', 'google-cloud-core': '2.3.2', 'google-cloud-storage': '2.7.0', 'google-crc32c': '1.5.0', 
'google-resumable-media': '2.4.1', 'googleapis-common-protos': '1.59.0', 'grpcio': '1.47.5', 'grpcio-tools': '1.47.5', 'gunicorn': '20.1.0', 'h11': '0.14.0', 'h2': '4.1.0', 'hpack': '4.0.0', 'httpcore': '0.16.3', 'httplib2': 
'0.19.1', 'httptools': '0.5.0', 'httpx': '0.23.3', 'humanize': '4.6.0', 'hyperframe': '6.0.1', 'idna': '3.4', 'importlib-metadata': '6.1.0', 'iniconfig': '2.0.0', 'ipykernel': '6.22.0', 'ipython': '8.11.0', 'ipython-genutils': 
'0.2.0', 'ipywidgets': '7.7.4', 'isodate': '0.6.1', 'itsdangerous': '2.1.2', 'jedi': '0.18.2', 'joblib': '1.2.0', 'jsonpointer': '2.3', 'jsonschema': '3.2.0', 'jupyter-client': '8.1.0', 'jupyter-core': '5.3.0', 'jupyter-events': 
'0.6.3', 'jupyter-server': '2.5.0', 'jupyter-server-terminals': '0.4.4', 'jupyterlab-pygments': '0.2.2', 'jupyterlab-widgets': '1.1.3', 'kfp': '1.8.16', 'kfp-pipeline-spec': '0.1.16', 'kfp-server-api': '1.8.5', 'kiwisolver': 
'1.4.4', 'kubernetes': '18.20.0', 'llvmlite': '0.39.1', 'lxml': '4.9.2', 'matplotlib': '3.7.1', 'matplotlib-inline': '0.1.6', 'mistune': '2.0.5', 'mlflow': '2.2.2', 'mmh3': '3.0.0', 'monotonic': '1.6', 'multidict': '6.0.4', 
'nbclassic': '0.5.3', 'nbclient': '0.7.2', 'nbconvert': '7.2.10', 'nbformat': '5.8.0', 'nest-asyncio': '1.5.6', 'notebook': '6.5.3', 'notebook-shim': '0.2.2', 'numba': '0.56.4', 'numpy': '1.24.2', 'oauthlib': '3.2.2', 'packaging': 
'23.0', 'pandas': '1.5.3', 'pandocfilters': '1.5.0', 'parso': '0.8.3', 'passlib': '1.7.4', 'pexpect': '4.8.0', 'pickleshare': '0.7.5', 'pip': '22.3.1', 'platformdirs': '3.2.0', 'pluggy': '1.0.0', 'prometheus-client': '0.16.0', 
'prompt-toolkit': '3.0.38', 'protobuf': '3.20.3', 'psutil': '5.9.4', 'ptyprocess': '0.7.0', 'pure-eval': '0.2.2', 'pyarrow': '11.0.0', 'pyasn1': '0.4.8', 'pyasn1-modules': '0.2.8', 'pycparser': '2.21', 'pydantic': '1.10.7', 
'pyg-nightly': '2.4.0.dev20230328', 'pymilvus': '2.2.3', 'pyparsing': '2.4.7', 'pyrsistent': '0.19.3', 'pytest': '7.2.2', 'python-dateutil': '2.8.2', 'python-dotenv': '1.0.0', 'python-jose': '3.3.0', 'python-json-logger': '2.0.7', 
'python-multipart': '0.0.6', 'python-terraform': '0.10.1', 'pytz': '2023.2', 'pyzmq': '25.0.2', 'qdrant-client': '0.11.10', 'querystring-parser': '1.2.4', 'rdflib': '6.3.2', 'requests': '2.28.2', 'requests-oauthlib': '1.3.1', 
'requests-toolbelt': '0.10.1', 'rfc3339-validator': '0.1.4', 'rfc3986': '1.5.0', 'rfc3986-validator': '0.1.1', 'rich': '12.6.0', 'rsa': '4.9', 'scikit-learn': '1.2.2', 'scipy': '1.10.1', 'setuptools': '67.6.1', 'shap': '0.41.0', 
'shellingham': '1.5.0.post1', 'six': '1.16.0', 'slicer': '0.0.7', 'smmap': '5.0.0', 'sniffio': '1.3.0', 'soupsieve': '2.4', 'sqlalchemy2-stubs': '0.0.2a32', 'sqlmodel': '0.0.8', 'sqlparse': '0.4.3', 'stack-data': '0.6.2', 
'starlette': '0.17.1', 'strip-hints': '0.1.10', 'tabulate': '0.9.0', 'termcolor': '2.2.0', 'terminado': '0.17.1', 'threadpoolctl': '3.1.0', 'tinycss2': '1.2.1', 'tomli': '2.0.1', 'torch': '1.13.1', 'torch-scatter': '2.1.0', 
'torch-sparse': '0.6.16', 'tornado': '6.2', 'tqdm': '4.65.0', 'traitlets': '5.9.0', 'typer': '0.7.0', 'typing-extensions': '4.5.0', 'ujson': '5.4.0', 'uritemplate': '3.0.1', 'urllib3': '1.26.15', 'uvicorn': '0.17.6', 'uvloop': 
'0.17.0', 'validators': '0.18.2', 'watchgod': '0.8.2', 'wcwidth': '0.2.6', 'webcolors': '1.13', 'webencodings': '0.5.1', 'websocket-client': '1.5.1', 'websockets': '10.4', 'wheel': '0.40.0', 'widgetsnbextension': '3.6.3', 'wrapt': 
'1.15.0', 'yarl': '1.8.2', 'zenml': '0.35.1', 'zipp': '3.15.0'}

### What happened?

While I was attempting to connect to a remote ZenML server in command line, a runtime error with a cryptic message occurred:
`RuntimeError: Error initializing rest store with URL 'https://foo.bar': Expecting value: line 1 column 1 (char 0)`

Turns out there was a HTTP 401 error when the client was communicating to the server. However, during the handling of it (see code below), the program assumes that the response body is in JSON format that's given by the ZenML service but in reality it actually was a piece of binary text from GCP proxy, thus there was an extra JSON decoder error and has given the above messages.

https://github.com/zenml-io/zenml/blob/5620f49e9e70af6b587dc21d6c181bbe42900c63/src/zenml/zen_stores/rest_zen_store.py#L1909-L1913

It would be nice if the response handler can handle this as well, thank you!

### Reproduction steps

1. Run `zenml connect  --url=https://foo.bar --username=******** --password=********` in command line

### Relevant log output

```shell
RuntimeError: Error initializing rest store with URL 'https://foo.bar': Expecting value: line 1 column 1 (char 0)
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by stefannica**
Fixes #1449

## Describe changes
We put this off for far too long: this PR formalizes error handling between the REST API server and client.

The PR also removes some exceptions that were rarely or never used.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Review by schustmi — COMMENTED** [review]
Still need to test this, but from a code perspective this looks good to me!

Can you think of some way to test this in the CI to make sure the conversion works both ways? Or maybe even just test the `http_exception_from_error` and `exception_from_response` interaction to make sure that is working as expected?

**Comment by stefannica** [comment]
> Still need to test this, but from a code perspective this looks good to me!
> 
> Can you think of some way to test this in the CI to make sure the conversion works both ways? Or maybe even just test the `http_exception_from_error` and `exception_from_response` interaction to make sure that is working as expected?

Yes, good point, I will just add some unit tests that target the conversion utilities.

**Review by fa9r — APPROVED** [review]
Yes!!! 🙏 🙏 🙏

File: src/zenml/zen_server/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    for exception, status_code in REST_API_EXCEPTIONS:
+        if isinstance(error, exception):
+            return HTTPException(
+                status_code=status_code, detail=error_detail(error)
+            )
+    else:
+        return HTTPException(status_code=500, detail=error_detail(error))
+
+
+def exception_from_response(
```
Comment (AlexejPenner) [inline src/zenml/zen_server/exceptions.py:122]: I think this could benefit from a small unit-test, just to ensure all corner cases are covered and future work around this does not silently break.

Comment (AlexejPenner) [reply src/zenml/zen_server/exceptions.py:122]: As I scanned up across the comments i just saw that you are still planning on writing unit tests. Feel free to disregard this then :)

File: src/zenml/zen_server/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    (ZenKeyError, 404),
+    (KeyError, 404),
+    # 400 Bad Request
+    (PydanticValidationError, 400),
+    (ValidationError, 400),
+    (ValueError, 400),
+    # 422 Unprocessable Entity
+    (NotImplementedError, 422),
+    (PydanticValidationError, 422),
+    (ValueError, 400),
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply src/zenml/zen_server/exceptions.py:82]: even then it wouldn't make sense to have two mappings for the Value Error

[the remark this answers is not in the document]
Comment (fa9r) [reply src/zenml/zen_server/exceptions.py:82]: It depends. It could still make sense. The current implementation will use the last entry as default error when mapping back, i.e., both 400 and 422 would now be mapped back to `ValueError` if the correct exception class cannot be determined. I guess this was the purpose of that entry.

File: src/zenml/zen_server/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    # 404 Not Found
+    (DoesNotExistException, 404),
+    (ZenKeyError, 404),
+    (KeyError, 404),
+    # 400 Bad Request
+    (PydanticValidationError, 400),
+    (ValidationError, 400),
+    (ValueError, 400),
+    # 422 Unprocessable Entity
+    (NotImplementedError, 422),
```
Comment (fa9r) [inline src/zenml/zen_server/exceptions.py:80]: As from our discussion, this should perhaps not map to 422 since it already maps to 501 below

File: src/zenml/zen_server/exceptions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/zen_server/exceptions.py:82]: This needs to be 422, good point, thanks for spotting it.

File: src/zenml/zen_server/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+# 2. When HTTP error response is received by the client, the error response is
+#  converted back to an exception based on the status code and the exception
+#  class name and arguments are extracted from the response body.
+#
+# NOTE: The order of the exceptions is important. The first matching
+# exception will be used, so the more specific exceptions should be
+# listed first. For error responses that cannot be unpacked into an
+# exception, the last (the more generic) exception in the list corresponding
+# to the status code will be used.
+REST_API_EXCEPTIONS: List[Tuple[Type[Exception], int]] = [
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/zen_server/exceptions.py:60]: Using a dictionary for the "exception to error code" specification wouldn't work, for one main reason: the lookup operation used to determine which error code to use for an exception is not equality (`==`) but `isinstance` instead. This allows a progression from more specific to less specific exception types, same you would do in a try-catch block with multiple errors. This also allows us to convert any future exception into an error code as long as it inherits from one of these as a base exception without having to add it explicitly to this list.

I also thought about splitting it into two maps, but it would duplicate a lot of information and make it more error-prone to maintain it in the future.

File: src/zenml/zen_server/exceptions.py (hunk 3 above)
Comment (stefannica) [reply src/zenml/zen_server/exceptions.py:80]: Correct ! I already removed this line.

**Review by fa9r — APPROVED** [review]
